internal/usde/usde.go

```diff
@@ -30,30 +31,43 @@ var (
 // - bool: Indicates whether the resource needs an update.
 // - apiv1.UpgradeStrategy: The most conservative upgrade strategy to be used for updating the resource.
 // - bool: Indicates if the controller managed resources should be recreated (delete-recreate).
+// - unstructured.UnstructuredList: Resources that require addition
+// - unstructured.UnstructuredList: Resources that require modification
+// - unstructured.UnstructuredList: Resources that require deletion
 // - error: Any error encountered during the function execution.
-func ResourceNeedsUpdating(ctx context.Context, newDef, existingDef *unstructured.Unstructured) (bool, apiv1.UpgradeStrategy, bool, error) {
+func ResourceNeedsUpdating(
+	ctx context.Context,
+	newDef, existingDef *unstructured.Unstructured,
+	newRiders []riders.Rider,
+	existingRiders unstructured.UnstructuredList) (bool, apiv1.UpgradeStrategy, bool, unstructured.UnstructuredList, unstructured.UnstructuredList, unstructured.UnstructuredList, error) {
 	numaLogger := logger.FromContext(ctx)
 
 	metadataNeedsUpdating, metadataUpgradeStrategy, err := resourceMetadataNeedsUpdating(ctx, newDef, existingDef)
 	if err != nil {
-		return false, apiv1.UpgradeStrategyError, false, err
+		return false, apiv1.UpgradeStrategyError, false, unstructured.UnstructuredList{}, unstructured.UnstructuredList{}, unstructured.UnstructuredList{}, err
 	}
 
 	specNeedsUpdating, specUpgradeStrategy, recreate, err := resourceSpecNeedsUpdating(ctx, newDef, existingDef)
 	if err != nil {
-		return false, apiv1.UpgradeStrategyError, false, err
+		return false, apiv1.UpgradeStrategyError, false, unstructured.UnstructuredList{}, unstructured.UnstructuredList{}, unstructured.UnstructuredList{}, err
+	}
+
+	ridersNeedUpdating, ridersUpgradeStrategy, additionsRequired, modificationsRequired, deletionsRequired, err := RidersNeedUpdating(ctx, existingDef.GetNamespace(), existingDef.GetKind(), existingDef.GetName(), newRiders, existingRiders)
+	if err != nil {
+		return false, apiv1.UpgradeStrategyError, false, additionsRequired, modificationsRequired, deletionsRequired, err
 	}
 
 	numaLogger.WithValues(
 		"metadataUpgradeStrategy", metadataUpgradeStrategy,
 		"specUpgradeStrategy", specUpgradeStrategy,
+		"ridersUpgradeStrategy", ridersUpgradeStrategy,
 	).Debug("upgrade strategies")
 
-	if !metadataNeedsUpdating && !specNeedsUpdating {
-		return false, apiv1.UpgradeStrategyNoOp, false, nil
+	if !metadataNeedsUpdating && !specNeedsUpdating && !ridersNeedUpdating {
+		return false, apiv1.UpgradeStrategyNoOp, false, additionsRequired, modificationsRequired, deletionsRequired, nil
 	}
 
-	return true, getMostConservativeStrategy([]apiv1.UpgradeStrategy{metadataUpgradeStrategy, specUpgradeStrategy}), recreate, nil
+	return true, getMostConservativeStrategy([]apiv1.UpgradeStrategy{metadataUpgradeStrategy, specUpgradeStrategy, ridersUpgradeStrategy}), recreate, additionsRequired, modificationsRequired, deletionsRequired, nil
```

Comment:
our USDE function now determines if any Rider has been added, modified, or deleted as well

if so, it either returns progressive or direct-apply depending on whether the Rider was indicated as `progressive=true`